dataactvalidator/validation_handlers/validationManager.py

```diff
@@ -318,16 +320,45 @@ def runCrossValidation(self, jobId, interfaces):
         failures.extend(Validator.crossValidateSql(rules, submissionId))
         bucketName = CONFIG_BROKER['aws_bucket']
         regionName = CONFIG_BROKER['aws_region']
-        errorFileName = self.getFileName(interfaces.jobDb.getCrossFileReportPath(submissionId))
         errorDb = interfaces.errorDb
 
-        with self.getWriter(regionName, bucketName, errorFileName,
-                            self.crossFileReportHeaders) as writer:
-            for failure in failures:
-                writer.write(failure)
-                errorDb.recordRowError(jobId, "cross_file",
-                    failure[0], failure[2], failure[4], failure[5])
-            writer.finishBatch()
+        # Get list of file types
+        fileTypes = interfaces.validationDb.getFileTypeList()
+        # Create dict mapping file types to empty lists
+        fileTypeDict = {targetFile:[] for targetFile in fileTypes}
+
+        # Sort failures by source and target file
+        failureDict = {sourceFile:copy.deepcopy(fileTypeDict) for sourceFile in fileTypes}
+
+        for failure in failures:
+            errorDb.recordRowError(jobId, "cross_file",
+                    failure[0], failure[2], failure[4], failure[5], failure[6], failure[7])
+            # Put failure into list for this source and target file
+            sourceType = interfaces.validationDb.getFileTypeById(failure[6])
+            targetType = interfaces.validationDb.getFileTypeById(failure[7])
+            if failure[7] < failure[6]:
+                # Lower ID goes first
+                failureDict[targetType][sourceType].append(failure)
+            elif failure[7] == failure[6]:
+                raise ValueError("Both files in cross file are the same type")
+            else:
+                failureDict[sourceType][targetType].append(failure)
+
+        # For each pair of files create an error report
+        for sourceFile in fileTypes:
```

Comment:
The fain/uri rule between C and D2 is applied in both directions (C8 and C9) .  So one error is that a fain/uri from file C was not found in D2, and the other was that a fain/uri in D2 was not found in C, this is why the source file is different for the two errors.  The displayed values come from the source file, and are being checked against the other file in the pair.